akka-sample-sharding-scala/killrweather/src/main/scala/sample/killrweather/WeatherRoutes.scala

```diff
@@ -2,64 +2,92 @@ package sample.killrweather
+    val ref = sharding.entityRefFor(WeatherStation.TypeKey, wsid)
+    ref.ask(WeatherStation.Query(wsid, dataType, function, _))
+  }
+
+  // unmarshallers for the query parameters
+  private val funcsFromName = WeatherStation.Function.All.map(function => function.toString.toLowerCase -> function).toMap
+  private implicit val functionTypeUnmarshaller = Unmarshaller.strict[String, WeatherStation.Function](text => funcsFromName(text.toLowerCase))
+
+  private val dataTypesFromNames = WeatherStation.DataType.All.map(dataType => dataType.toString.toLowerCase -> dataType).toMap
+  private implicit val dataTypeUnmarshaller = Unmarshaller.strict[String, WeatherStation.DataType](text => dataTypesFromNames(text.toLowerCase))
+
+
+  // imports needed for the routes and entity json marshalling
   import akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport._
   import akka.http.scaladsl.server.Directives._
   import JsonFormats._
 
   val weather: Route =
     pathPrefix("weather") {
       concat(
-        post {
-          entity(as[Aggregator.Data]) { data =>
-            val f: Future[WeatherRoutes.DataIngested] = guardian.ask(Guardian.Ingest(data.wsid, data, _))
-            onSuccess(f) { performed =>
-              complete(StatusCodes.Accepted -> s"$performed from event time: ${data.eventTime}")
+        pathPrefix(Segment) { wsid =>
+          concat(
+            pathEnd {
+              concat(
+                get {
+                  parameters(("type".as[WeatherStation.DataType], "function".as[WeatherStation.Function])) { (dataType, function) =>
+                    complete(query(wsid, dataType, function))
+                  }
+                },
+                post {
+                  onSuccess(addWeatherStation(wsid)) { performed =>
+                    complete(StatusCodes.Created -> performed.wsid)
+                  }
+                }
+              )
+            },
+            path("data") {
+              post {
+                entity(as[WeatherStation.Data]) { data =>
+                  onSuccess(recordData(wsid, data)) { performed =>
+                    complete(StatusCodes.Accepted -> s"$performed from event time: ${data.eventTime}")
+                  }
+                }
+              }
             }
-          }
+          )
         },
-        (path(Segment) & (put | post)) { wsid =>
-          val f: Future[WeatherRoutes.WeatherStationAdded] = guardian.ask(replyTo => Guardian.AddWeatherStation(wsid, replyTo))
-          onSuccess(f) { performed =>
-            complete(StatusCodes.Created -> performed.wsid)
-          }
+        pathEnd {
+          concat(
+            get {
+              complete(stations.ask(Stations.GetWeatherStations))
+            }
+          )
         },
-        get {
-          onSuccess(guardian.ask(Guardian.GetWeatherStations)) { extraction =>
-            complete(StatusCodes.Created -> extraction)
-          }
-        })
-      // todo curl queries
+       )
```

Comment:
Restructured the routes to be more clear and potentially more rest:y

Reply:
Bueno.